pythran/pythonic/numpy/dot.hpp

```diff
@@ -1,60 +1,189 @@
+                    return dot(e_, f_);
                 }
 
+        // If arguments could be use with blas, we evaluate them as we need pointer on array for blas
         template<class E, class F>
-            typename std::enable_if<
-            (std::is_scalar<E>::value or types::is_complex<E>::value) and
-            types::is_numexpr_arg<F>::value and types::numpy_expr_to_ndarray<F>::N ==1,
-            decltype(std::declval<E>()*std::declval<typename types::numpy_expr_to_ndarray<F>::T>())
-                >::type
-                dot(F const& f,types::list<E> const& e) {
-                    return dot(f, asarray(e));
+            typename std::enable_if<types::is_numexpr_arg<E>::value and types::is_numexpr_arg<F>::value // It is an array_like
+                                    and not (types::is_ndarray<E>::value and types::is_ndarray<F>::value)
+                                    and is_blas_type<typename E::dtype>::value and is_blas_type<typename F::dtype>::value // With dtype compatible with blas
+                                    and E::value == 1 and F::value == 2, // And it is vect / matrix
+                                    types::ndarray<typename __combined<typename E::dtype, typename F::dtype>::type, 1>
+                                   >::type
+                dot(E const& e, F const& f) {
+                    types::ndarray<typename __combined<typename E::dtype, typename F::dtype>::type, 1> e_ = e;
+                    types::ndarray<typename __combined<typename E::dtype, typename F::dtype>::type, 2> f_ = f;
+                    return dot(e_, f_);
                 }
 
+        // If one of the arg doesn't have a "blas compatible type", we use a slow matrix vector multiplication.
         template<class E, class F>
-            typename std::enable_if<
-            (std::is_scalar<E>::value or types::is_complex<E>::value) and
-            types::is_numexpr_arg<F>::value and types::numpy_expr_to_ndarray<F>::N ==1,
-            decltype(std::declval<E>()*std::declval<typename types::numpy_expr_to_ndarray<F>::T>())
-                >::type
-                dot(types::list<E> const& e, F const& f) {
-                    return dot(asarray(e), f);
+            typename std::enable_if<(not is_blas_type<typename E::dtype>::value or not is_blas_type<typename F::dtype>::value)
+                                    and E::value == 1 and F::value == 2, // And it is vect / matrix
+                                    types::ndarray<typename __combined<typename E::dtype, typename F::dtype>::type, 1>
+                                    >::type
+            dot(E const& e, F const& f)
+            {
+                types::ndarray<typename __combined<typename E::dtype, typename F::dtype>::type, 1> out(types::array<long, 1>{{f.shape[1]}}, 0);
+                for(long i=0; i<out.shape[0]; i++)
+                    for(long j=0; j<f.shape[0]; j++)
+                        out[i] += e[j] * f[types::array<long, 2>{{j, i}}];
```

Comment:
As I had some issue with gexpr (uncorrect results), I prefer keep it that way until you PR is merged.